When updating faerie I get a linker error on linux

```
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/home/bjorn/Documenten/rustc_codegen_cranelift/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/std.dummy_name.rcgu.o" "-o" "/home/bjorn/Documenten/rustc_codegen_cranelift/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/libstd.so" "/home/bjorn/Documenten/rustc_codegen_cranelift/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps/std.allocator_shim.rcgu.o" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs" "-L" "/home/bjorn/Documenten/rustc_codegen_cranelift/build_sysroot/target/x86_64-unknown-linux-gnu/debug/deps" "-L" "/home/bjorn/Documenten/rustc_codegen_cranelift/build_sysroot/target/debug/deps" "-L" "sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-ldl" "-lrt" "-lpthread" "-Wl,-Bstatic" "-Wl,--whole-archive" "/tmp/rustc5pSiro/libpanic_abort-ebe895f026b5ad95.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc5pSiro/libunwind-9bf4ad135fcef4eb.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc5pSiro/librustc_demangle-47343bf5c05ad1ab.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc5pSiro/liblibc-46063ca8474f398d.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc5pSiro/librustc_std_workspace_core-96d00f6d22747389.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc5pSiro/liballoc-8684167332e0bd28.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc5pSiro/libcore-1958aff3d7d300b8.rlib" "-Wl,--no-whole-archive" "/tmp/rustc5pSiro/libcompiler_builtins-2fce88ecd3031979.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil" "-shared"
  = note: /tmp/rustc5pSiro/libpanic_abort-ebe895f026b5ad95.rlib: member /tmp/rustc5pSiro/libpanic_abort-ebe895f026b5ad95.rlib(dummy_namercgu) in archive is not an object
          collect2: error: ld returned 1 exit status
```

Repro at https://github.com/bjorn3/rustc_codegen_cranelift/tree/link_error

Got another error:

```
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/bjorn/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "target/out/arbitrary_self_types_pointers_and_wrappers.dummy_name.rcgu.o" "-o" "target/out/arbitrary_self_types_pointers_and_wrappers" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs" "-L" "target/out" "-L" "/home/bjorn/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/home/bjorn/Documenten/rustc_codegen_cranelift/target/out/libmini_core.rlib" "-Wl,-Bdynamic" "-lc"
  = note: /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/Scrt1.o: In function `_start':
          (.text+0x20): undefined reference to `main'
          collect2: error: ld returned 1 exit status
```

According to `readelf` `main` does exist.

Repro: bjorn3/rustc_codegen_cranelift@b202dd7bed9371afa22508e18b595ca876218580

--- a/src/elf.rs
+++ b/src/elf.rs
@@ -469,7 +469,6 @@ impl<'a> Elf<'a> {
         }
     }
     pub fn add_definition(&mut self, name: &str, data: &'a [u8], decl: &artifact::DefinedDecl) {
-        // sh_info requires nsections + nlocals to add as delimiter; see the associated FunFact
         let def_size = data.len();
 
         let section_name = match decl {
@@ -530,8 +529,9 @@ impl<'a> Elf<'a> {
                     .create();
                 // insert it into our symbol table
                 self.symbols.insert(idx, symbol);
+                // sh_info requires nsections + nlocals to add as delimiter; see the associated FunFact
                 // nonglobals go into the symbol table first (per iteration through definitions in
-                // caller). we need to track how many locals there are
+                // caller)
                 if !decl.is_global() {
                     self.nlocals += 1;
                 }

--- a/src/elf.rs
+++ b/src/elf.rs
@@ -470,9 +470,6 @@ impl<'a> Elf<'a> {
     }
     pub fn add_definition(&mut self, name: &str, data: &'a [u8], decl: &artifact::DefinedDecl) {
         // sh_info requires nsections + nlocals to add as delimiter; see the associated FunFact
-        if !decl.is_global() {
-            self.nlocals += 1;
-        }
         let def_size = data.len();
 
         let section_name = match decl {
@@ -533,6 +530,11 @@ impl<'a> Elf<'a> {
                     .create();
                 // insert it into our symbol table
                 self.symbols.insert(idx, symbol);
+                // nonglobals go into the symbol table first (per iteration through definitions in
+                // caller). we need to track how many locals there are
+                if !decl.is_global() {
+                    self.nlocals += 1;
+                }
             }
             DefinedDecl::DebugSection(_) => {
                 // No symbols in debug sections, yet...

